2015-Fall/adrianbrink/SmashTag/SmashTag/Tweet.swift

```diff
@@ -0,0 +1,159 @@
+            for mediaData in mediaEntities {
+                if let mediaItem = MediaItem(data: mediaData as? NSDictionary) {
+                    accumulatedMedia.append(mediaItem)
+                }
+            }
+        }
+        media = accumulatedMedia
+        let hashtagMentionsArray = data?.valueForKeyPath(TwitterKey.Entities.Hashtags) as? NSArray
+        hashtags = Tweet.getIndexedKeywords(hashtagMentionsArray, inText: text, prefix: "#")
+        let urlMentionsArray = data?.valueForKeyPath(TwitterKey.Entities.URLs) as? NSArray
+        urls = Tweet.getIndexedKeywords(urlMentionsArray, inText: text, prefix: "h")
+        let userMentionsArray = data?.valueForKeyPath(TwitterKey.Entities.UserMentions) as? NSArray
+        userMentions = Tweet.getIndexedKeywords(userMentionsArray, inText: text, prefix: "@")
+        
+        if user == nil || text == nil || created == nil {
+            return nil
+        }
+    }
+
+    private class func getIndexedKeywords(dictionary: NSArray?, inText: String, prefix: String? = nil) -> [IndexedKeyword] {
+        var results = [IndexedKeyword]()
+        if let indexedKeywords = dictionary {
+            for indexedKeywordData in indexedKeywords {
+                if let indexedKeyword = IndexedKeyword(data: indexedKeywordData as? NSDictionary, inText: inText, prefix: prefix) {
+                    results.append(indexedKeyword)
+                }
+            }
+        }
+        return results
+    }
+    
+    struct TwitterKey {
+        static let User = "user"
+        static let Text = "text"
+        static let Created = "created_at"
+        static let ID = "id_str"
+        static let Media = "entities.media"
+        struct Entities {
+            static let Hashtags = "entities.hashtags"
+            static let URLs = "entities.urls"
+            static let UserMentions = "entities.user_mentions"
+            static let Indices = "indices"
+        }
+    }
+}
+
+private extension NSString {
+    func rangeOfString(substring: NSString, nearRange: NSRange) -> NSRange {
+        var start = max(min(nearRange.location, length-1), 0)
+        var end = max(min(nearRange.location + nearRange.length, length), 0)
+        var done = false
+        while !done {
+            let range = self.rangeOfString(substring as String, options: NSStringCompareOptions(), range: NSMakeRange(start, end-start))
+            if range.location != NSNotFound {
+                return range
+            }
+            done = true
+            if start > 0 { start-- ; done = false }
+            if end < length { end++ ; done = false }
+        }
+        return NSMakeRange(NSNotFound, 0)
+    }
+}
+
+private extension String {
+    var asTwitterDate: NSDate? {
+        get {
+            let dateFormatter = NSDateFormatter()
+            dateFormatter.dateFormat = "EEE MMM dd HH:mm:ss Z yyyy"
+            return dateFormatter.dateFromString(self)
```

Comment:
On my device this did not work because my locale was in `da_DK`. This meant the `created` property of the `Tweet` was nil. I fixed it by adding a locale explicitly:

    dateFormatter.locale = NSLocale(localeIdentifier: "us")


To explain this. `NSDate` is always in GMT. Displaying or formatting dates is w.r.t. a locale, but this is never set on `NSDate`. `NSDateFormatter` takes your device's locale behind the scenes and uses that to format a `NSDate` for display or parsing. As I do above you can override the locale used by `NSDateFormatter`. This is, for example, also how `NSDatePicker` works.